### PR TITLE
Initial Implementation of MediaTek Backend for Executorch

### PR DESCRIPTION
**Pull request by neuropilot-captain**
This pull request introduces the initial implementation of the MediaTek backend for Executorch. Below are the key points and instructions related to this contribution:

- "Ahead of Time" and "Runtime" are included in this implementation.
- The build process and execution phase rely on libraries provided by the MediaTek NeuroPilot SDK. For the purpose of verification, these libraries will be made available offline.

### Build Instructions:
To build the MediaTek backend for Executorch, please adhere to the guidelines provided in the `backends/mediatek/scripts/README.md` file within this repository.

### Ahead-Of-Time Compilation
To compile models for MediaTek devices, please follow the instruction on the ExecuTorch documentation.
For Llama compilation, please refer to `examples/mediatek/README.md` for the usage of the export script `shell_scripts/export_llama.sh`

### On-device Execution:
For instructions on how to run an example on MediaTek devices, refer to the documentation in `examples/mediatek/README.md`.

### Llama Example:
A sample Llama runner is built together with MediaTek backend.

### CODE REVIEW

File: backends/mediatek/scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Build Step
+1. Download **Buck2**
+Download buck2 from https://github.com/facebook/buck2/releases/tag/2023-07-18, and add the directory to $PATH
+
+2. Download **Android NDK**
+Download Android NDK and make sure that $ANDROID_NDK is specified as the NDK Path.
+
+3. Download **MediaTek ExercuTorch libraries**
+Download related libraries.
```
Comment (kimishpatel) [inline backends/mediatek/scripts/README.md:9]: What are these and where do you get it from? I would like to repro

File: examples/mediatek/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# ExecuTorch Neuron Backend examples
+
+## Directory structure
```
Comment (kimishpatel) [inline examples/mediatek/README.md:3]: Can you add pre-requisite here as to which phones support MTK chipset? ALso do add "validated on" to indicate which phone or chipset this is validated on

File: examples/mediatek/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+├── CMakeLists.txt                    # The build file
+└── README.md                         # This file
+```
+
+## Environment
+
+Please follow the build step in backends/mediatek/scripts.
+The mtk_executor_runner will be automatically generated.
+
+## Execute
```
Comment (kimishpatel) [inline examples/mediatek/README.md:16]: How do you generate the pte file?

**Comment by kimishpatel** [comment]
I dont see any instructions/code on the ahead of time component

**Comment by mergennachin** [comment]
Thanks for your contribution and PR.

Will take a look at it deeply soon. I just kicked off OSS CI jobs. One thing, we have a lint rule. Here's how you can invoke:

https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#coding-style

**Comment by mergennachin** [comment]
> The build process and execution phase rely on libraries provided by the MediaTek NeuroPilot SDK. For the purpose of verification, these libraries will be made available offline.

Is there a public link that we can download from MediaTek website? If not, is there a plan to publish it publicly?

**Review by swolchok — COMMENTED** [review]
just dropping in with a bunch of small C++ suggestions; feel free to apply them or not

File: examples/mediatek/executor_runner/mtk_executor_runner.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  // be used by a single thread at at time, but it can be reused.
+  //
+
+  Result<Method> method = program->load_method(method_name, &memory_manager);
+  ET_CHECK_MSG(
+      method.ok(),
+      "Loading of method %s failed with status 0x%" PRIx32,
+      method_name,
+      (uint32_t)method.error());
+  ET_LOG(Info, "Method loaded.");
```
Comment (shoumikhin) [inline examples/mediatek/executor_runner/mtk_executor_runner.cpp:155]: Using the module extension we can skip most of this loading code:

```
#include <executorch/extension/module/module.h>

// ...

Then use it as:

const auto result = module.forward(inputs);

```suggestion
```

**Review by cccclai — COMMENTED** [review]
Thank you pushing the first PR! Looks great :)

Overall I think `NeuronBackend` and `mtk_executor_runner` are reasonable. It will be great if we can share the expected flow for users to generate the .pte file and run it.

Regarding the llama runner logic, I can see why we need the multimodel loader and the caching logic, but it's not quite clear to me what logic from `.so`is needed here. Maybe we can walk through it.

File: examples/mediatek/executor_runner/mtk_executor_runner.cpp (hunk 4 above)
Comment (cccclai) [reply examples/mediatek/executor_runner/mtk_executor_runner.cpp:155]: I'd be careful with this - the current default `module` has some default logic, we'd need to visit if they meet the need or may regress the perf.

File: backends/mediatek/runtime/NeuronExecutor.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    err |= NeuronCompilation_setPriority(compilation, NEURON_PRIORITY_HIGH);
+    CHECK_NO_ERROR(err);
+
+    err = NeuronCompilation_finish(compilation);
+    CHECK_NO_ERROR(err);
+
+
+    // ---------------------------Execution------------------------------------
+    // Run the compiled model against a set of inputs
+    err = NeuronExecution_create(compilation, &execution);
```
[the remark this answers is not in the document]
Comment (neuropilot-captain) [reply backends/mediatek/runtime/NeuronExecutor.cpp:149]: No, this function only creates an execution instance for `execution`, not executing it.

File: backends/mediatek/runtime/NeuronBackend.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            mExecutor.SetInputOutputFromMemory</*isInput*/false>(output_index, unit->GetNeuronMemory(), offset, data_size);
+        } else {
+            mExecutor.SetInputOutput</*isInput=*/false>(output_index, data_ptr, data_size);
+        }
+    }
+
+    return mExecutor.Compute() == NEURON_NO_ERROR ? Error::Ok : Error::InvalidState;
+};
+
+int NeuronExecuTorchDelegate::HintNeuronBackend(EValue** args) const {
```
[the remark this answers is not in the document]
Comment (neuropilot-captain) [reply backends/mediatek/runtime/NeuronBackend.cpp:117]: 'import forever' is an optimization hint designed for specific models including Llama. We prefer not to automatically apply it to every model because it could leads unknow effects.

File: examples/mediatek/executor_runner/mtk_executor_runner.cpp (hunk 4 above)
Comment (neuropilot-captain) [reply examples/mediatek/executor_runner/mtk_executor_runner.cpp:155]: The current implementation of `mtk_executor_runner` is based on `examples/portable/executor_runner`, and it has passed testing without any issues.
We will look into `executorch/extension/module` to determine if we need to make any modifications to our code in the future.

File: examples/mediatek/executor_runner/mtk_executor_runner.cpp [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  // memory-planning pasees.
+  //
+  // Each buffer typically corresponds to a different hardware memory bank. Most
+  // mobile environments will only have a single buffer. Some embedded
+  // environments may have more than one for, e.g., slow/large DRAM and
+  // fast/small SRAM, or for memory associated with particular cores.
+  std::vector<std::unique_ptr<uint8_t[]>> planned_buffers; // Owns the memory
+  std::vector<Span<uint8_t>> planned_spans; // Passed to the allocator
+  size_t num_memory_planned_buffers = method_meta->num_memory_planned_buffers();
+  for (size_t id = 0; id < num_memory_planned_buffers; ++id) {
```
[the remark this answers is not in the document]
Comment (neuropilot-captain) [reply examples/mediatek/executor_runner/mtk_executor_runner.cpp:128]: As answered in the conversation below.
The current implementation of `mtk_executor_runner` is based on `examples/portable/executor_runner,` and we'll look into `executorch/extension/module `to determine if we need to modify the code.

File: examples/mediatek/README.md (hunk 2 above)
Comment (neuropilot-captain) [reply examples/mediatek/README.md]: Thanks for your suggestion! We have added the list of supported chips that the examples are validated on. Also, we have refactor this README for better readability.

File: backends/mediatek/scripts/README.md [hunk 8]
```diff
@@ -1,13 +1,35 @@
...
+- **Add to PATH**: Extract the downloaded file and add the directory to your system's `$PATH` environment variable.
+
+### 2. Android NDK
+
+- **Download Android NDK**: Acquire the Android NDK from the [Android developer site](https://developer.android.com/ndk/downloads).
+- **Set NDK Path**: Ensure that the `$ANDROID_NDK` environment variable is set to the path where the NDK is located.
+
+### 3. MediaTek ExercuTorch Libraries
+
+Download the following libraries required for MediaTek ExercuTorch:
```
Comment (cccclai) [inline backends/mediatek/scripts/README.md:21]: > Download the following libraries required ...

rephrase it to "Download the following libraries required for MediaTek libraries from xx portal (linked to be added)" just so it's clearer where it comes from

File: backends/mediatek/runtime/include/NeuronBackend.h [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+      return "--apusys-config \"{ \\\"high_addr\\\": true, \\\"import_forever\\\": true }\"";
+    } else if (mHighAddr) {
+      return "--apusys-config \"{ \\\"high_addr\\\": true }\"";
+    } else if (mImportForever) {
+      return "--apusys-config \"{ \\\"import_forever\\\": true }\"";
+    } else {
+      return "";
+    }
+  }
+};
```
[the remark this answers is not in the document]
Comment (neuropilot-captain) [reply backends/mediatek/runtime/include/NeuronBackend.h:65]: We appreciate your suggestion to return a const char*, as it's only used to call c_str(). However, we've opted for std::string to maintain flexibility for potential future enhancements.

File: backends/mediatek/runtime/include/NeuronBackend.h [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+      __ET_UNUSED BackendExecutionContext& context,
+      DelegateHandle* handle,
+      EValue** args) const override;
+
+  void destroy(DelegateHandle* handle) const override;
+
+  bool is_available() const override;
+};
+
+constexpr char kHighAddrKey[] = "HighAddr";
```
[the remark this answers is not in the document]
Comment (neuropilot-captain) [reply backends/mediatek/runtime/include/NeuronBackend.h]: Thank you for your feedback. I have moved the definition to the .cpp file and placed it in an anonymous namespace as suggested.

File: backends/mediatek/runtime/NeuronExecutor.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (cccclai) [reply backends/mediatek/runtime/NeuronExecutor.cpp:148]: The comment said "Run the compiled model against a set of inputs", and I thought it was executing the model. If not, can we update the comment to make it clear?

File: backends/mediatek/runtime/NeuronBackend.cpp [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+            LogInfo("NeuronBackend", "IsImportForever Enable : %d", setting.mImportForever);
+        } else {
+            LogWarn("NeuronBackend", "unknown compile spec: %s", compile_spec.key);
+        }
+    }
+    auto Payload = NeuronPayload(processed->data(), processed->size());
+    LogInfo("NeuronBackend", "version %u, input %u, output %u, length %u, payload size: %zu",
+                 Payload.Header.Version, Payload.Header.InputCount, Payload.Header.OutputCount, Payload.Header.DataLen, processed->size());
+
+    auto delegate = std::unique_ptr<NeuronExecuTorchDelegate>(
```
[the remark this answers is not in the document]
Comment (neuropilot-captain) [reply backends/mediatek/runtime/NeuronBackend.cpp:49]: In our design, a NeuronExecuTorchDelegate handles a **partition** of the model on the NPU. This separation ensures modularity and flexibility, making the codebase cleaner and easier to maintain.

File: backends/mediatek/runtime/NeuronBackend.cpp [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+Error NeuronExecuTorchDelegate::execute(
+      __ET_UNUSED BackendExecutionContext& context,
+      EValue** args) const {
+    if (HintNeuronBackend(args) != NEURON_NO_ERROR) {
+        return Error::InvalidState;
+    };
+
+    auto& allocator = GET_NEURON_ALLOCATOR;
```
[the remark this answers is not in the document]
Comment (neuropilot-captain) [reply backends/mediatek/runtime/NeuronBackend.cpp:81]: I apologize if I misunderstood your question. The NEURON_ALLOCATOR is a global instance that allows users to set up shared memory, which can then be utilized by the NeuronBackend.

File: backends/mediatek/runtime/NeuronBackend.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (cccclai) [reply backends/mediatek/runtime/NeuronBackend.cpp:120]: I see - hmm probably add more comments about it (or maybe just copy paste your response)

File: backends/mediatek/runtime/NeuronBackend.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (neuropilot-captain) [reply backends/mediatek/runtime/NeuronBackend.cpp:81]: Thank you for your response and suggestions.

We understand that we can customize the runtime allocator and point the temporary allocator to `GET_NEURON_ALLOCATOR`, so we don't need to use `NEURON_ALLOCATOR` as a global instance. This is indeed a good solution.

However, in our implementation, `NeuronBackend` needs to determine which device's buffer an address belongs to. Therefore, we use an additional API `Find` from `NEURON_ALLOCATOR` to query the backend. The current API in `BackendExecutionContext` does not allow us to achieve this functionality.

File: backends/mediatek/runtime/NeuronBackend.cpp [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+Error NeuronBackend::execute(__ET_UNUSED BackendExecutionContext& context,
+                             DelegateHandle* handle,
+                             EValue** args) const {
+    NeuronExecuTorchDelegate* delegate = reinterpret_cast<NeuronExecuTorchDelegate*>(handle);
+    return delegate->execute(context, args);
+}
+
+void NeuronBackend::destroy(DelegateHandle* handle) const {
+    NeuronExecuTorchDelegate* delegate = reinterpret_cast<NeuronExecuTorchDelegate*>(handle);
+    delete delegate;
```
Comment (mergennachin) [inline backends/mediatek/runtime/NeuronBackend.cpp:67]: you might have to call the destructor directly. Something like this. `delegate->~NeuronExecuTorchDelegate()`

File: examples/mediatek/model_export_scripts/llama.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+def get_argument_parser():
+    parser = argparse.ArgumentParser(
+        description='Run Export to ET for suppoorted LLM models.', allow_abbrev=False)
+    parser.add_argument('config', type=str,
+        help="[Required] Model config json file. "
+        "Model config must be in same directory as all model weight bins and tokenizer files.")
+    parser.add_argument('-p', '--precision', type=str, default='A16W8', choices=['A16W4', 'A16W8', 'A16W16', 'A8W4', 'A8W8'],
+        help="Precision to quantize entire model to.")
+    parser.add_argument('-n', '--num_chunks', type=int, default=4,
```
Comment (cccclai) [inline examples/mediatek/model_export_scripts/llama.py:39]: If it's stories model (the same model arch as llama2, 125M parameters), is the flow going to work? I believe in this case, we don't need to chunk the model: 
model checkpoint: https://huggingface.co/karpathy/tinyllamas/resolve/main/stories110M.pt
tokenizer: https://raw.githubusercontent.com/karpathy/llama2.c/master/tokenizer.model

File: examples/mediatek/models/llm_models/weights/.gitignore [hunk 15]
```diff
@@ -0,0 +1,2 @@
+*
```
Comment (cccclai) [inline examples/mediatek/models/llm_models/weights/.gitignore:1]: Is it a wip folder?

File: examples/mediatek/models/llm_models/modeling_common.py [hunk 16]
```diff
@@ -0,0 +1,497 @@
...
+        head_dim = int(self.config.hidden_size/self.config.num_attention_heads)
+        example_inputs = (
+            torch.randn(1, num_token, self.config.hidden_size, device='cpu', dtype=torch.float32),
+            torch.randn(1, 1, num_token, cache_size+num_token, device='cpu', dtype=torch.float32),
+            torch.randn(1, 2, num_token, head_dim, device='cpu', dtype=torch.float32),
+            *[torch.randn(1, self.config.num_key_value_heads, cache_size, head_dim, device='cpu', dtype=torch.float32) for _ in range(2*self.num_blocks)]
+        )
+        # Note: Assume cache size fixed shape as torch dynamic shape cannot handle dim 3 being
+        # combination of 2 dynamic dims
+        if get_dym_shape:
```
Comment (cccclai) [inline examples/mediatek/models/llm_models/modeling_common.py:486]: Are we using dynamic shape?

File: examples/mediatek/models/llm_models/modeling_llama.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+class LlamaModelChunk(ModelChunk):
+    def __init__(
+        self,
+        config: LlamaConfig,
+        num_blocks,
+        chunk_idx,
+        dtype=torch.float32,
+        include_tail=False,
+        return_attn=False,
+        jit_trace=False,
```
Comment (cccclai) [inline examples/mediatek/models/llm_models/modeling_llama.py:65]: what does `jit_trace` mean?

**Review by cccclai — COMMENTED** [review]
Thank you for putting up the AOT part prs! It's a long pr and I'm just half on it. Will continue reviewing later.

File: examples/mediatek/model_export_scripts/llama.py (hunk 14 above)
Comment (neuropilot-captain) [reply examples/mediatek/model_export_scripts/llama.py:39]: Hi @cccclai , generally, if its <= 3B param models we can do 1 chunk (ie effectively no chunking). 

Users just need to set -n 1 in the export shell script. 

For >= 7B params we need minimum 2 chunks.

File: examples/mediatek/models/llm_models/modeling_common.py (hunk 16 above)
Comment (neuropilot-captain) [reply examples/mediatek/models/llm_models/modeling_common.py:486]: Dynamic Shape is more for the case where users would like to calibrate on their own dataset. 

With dynamic shape, users can avoid padding each input to the length. 

Also, it is only used during the `capture_pre_autograd_graph`  part of the export flow. 

Subsequently the shape is fixed during the `torch.export.export` part

File: examples/mediatek/models/llm_models/weights/.gitignore (hunk 15 above)
Comment (neuropilot-captain) [reply examples/mediatek/models/llm_models/weights/.gitignore:1]: Mainly just to indicate to users where to put their model weights in.

File: examples/mediatek/models/llm_models/modeling_llama.py (hunk 17 above)
Comment (neuropilot-captain) [reply examples/mediatek/models/llm_models/modeling_llama.py:65]: It's a flag to skip all the forced float casting and some assertion checks.

File: examples/mediatek/model_export_scripts/llama.py [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+
+def export_to_et_ir(
+    output_folder, exp_name, model, precision, max_num_token, max_cache_size, chunk_idx, export_shapes
+):
+    print(f"Exporting Chunk {chunk_idx} to PTE")
+    example_inputs, dynamic_shapes = model.get_example_inputs(max_num_token, max_cache_size, True)
+
+    print("Getting pre autograd ATen Dialect Graph")
+    pre_autograd_aten_dialect = torch._export.capture_pre_autograd_graph(model, example_inputs, dynamic_shapes=dynamic_shapes) # NOTE: Will be replaced with export
+    quantizer = NeuropilotQuantizer()
```
Comment (cccclai) [inline examples/mediatek/model_export_scripts/llama.py:75]: Is `NeuropilotQuantizer` hosted on NP portal or github?

File: backends/mediatek/neuropilot/preprocess.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        num_outputs = len(edge_program.graph_signature.user_outputs)
+
+        compile_options = ['--arch=mdla5.1', '--relax-fp32']
+        for spec in module_compile_spec:
+            if spec.key in SKIP_COMPILE_SPEC_KEYS:
+                continue
+            value = spec.value.decode('utf-8')
+            compile_options.append(f'--{spec.key}={value}')
+
+        converter = mtk_converter.PyTorchV2Converter.from_exported_program(edge_program)
```
Comment (cccclai) [inline backends/mediatek/neuropilot/preprocess.py:52]: It seems like the `mtk_convert` hosted on NP portal is handling go through the graph and process node by node.

From a user perspective, without knowing what ops can be supported by mtk is a bit hard to lower a model to mtk backend. There can be two scenarios:
1. Partially lower the model. While it is the plan to partially lower some oss models to npu, it's a bit hard to tell if it will be supported smoothly as everything is hidden behind `mkt_convert`. 
2. Decompose the not supported ops to the op list supported by npu to fully lower a model to mtk. Similar to scenario 2 (also similar the `decompose_scaled_dot_product_attention` pass), users may want to decompose some non supported ops with the supported ops. It's hard to do so without know what ops are supported.

File: backends/mediatek/neuropilot/quantizer/annotator.py [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+    _mark_as_annotated([node, weight_node])
+
+
+@register_annotator(
+    [
+        torch.ops.aten.add.Scalar,
+        torch.ops.aten.add.Tensor,
+        torch.ops.aten.add_.Scalar,
+        torch.ops.aten.add_.Tensor,
+        torch.ops.aten.bmm.default,
```
Comment (cccclai) [inline backends/mediatek/neuropilot/quantizer/annotator.py:220]: I asked this question in the meeting and actually this part of code answered my question. The quantizer looks reasonable to me now.

**Review by cccclai — COMMENTED** [review]
I just finish reading through - the process in `examples/mediatek/model_export_scripts/llama.py` looks neat. Thank you for sending the high quality code.

One question I have is mostly regarding the supported ops in the backend.preprocess. Mostly around what ops can be supported by users and how to allow users to figure out by themselves to either partially lower their models to npu or fully lower to npu.

File: backends/mediatek/neuropilot/quantizer/annotator.py (hunk 20 above)
Comment (cccclai) [reply backends/mediatek/neuropilot/quantizer/annotator.py:220]: Yeah, it's more like describing how the ops are quantized.

File: examples/mediatek/README.md
Comment (Riandy) [inline examples/mediatek/README.md:1]: - Add that only linux environment is supported (this is due to the extra libs mtk_neuron and mtk_converter whl files are only supported in linux)
- Add hyperlink to ET environment setup https://pytorch.org/executorch/stable/getting-started-setup.html  
- Mention that requirements.txt file has two dependencies that cannot be installed via pip. 
  - mtk_converter
  - mtk-neuron
  - Another option is to split this up by removing those two from requirements.txt, so devs can just `pip3 install -r requirements.txt` and then explicitly mention that they need to install the other two libs.

```
// Ensure that you are inside executorch/examples/mediatek directory
pip3 install -r requirements.txt

// Download the two whl files from (some_location/URL)
pip3 install mtk_neuron-8.2.2-py3-none-linux_x86_64.whl
pip3 install mtk_converter-8.8.0.dev20240723+public.d1467db9-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
```
- Explain that dev should expect 8 pte files (4 for prompt and 4 for generation) + 1 token embedding file and provide location of the token embedding file location 
 ` models/llm_models/weights/<model_name>/<embedding_file_name>.bin `
- For llama models, need to use `mtk_llama_executor_runner`, and ask them to check `examples/mediatek/executor_runner/run_llama3_sample.sh` for reference

File: backends/mediatek/scripts/README.md
Comment (Riandy) [inline backends/mediatek/scripts/README.md:1]: - Buck2 Build Tool
  - This seems to be not needed? Looks like ET will auto download the required version it needs? (See: `executorch/build/resolve_buck.py`)
- Android NDK
  - Any specific version that is needed? Or tested upon? Maybe we should specify the version tested?
- “Build Script” section
  - Mentioned the output of the build script. Which is in the “`cmake-android-out`” folder.
In this case, it looks like we have 3 runners generated (executorch_runner, mtk_executor_runner and `mtk_llama_executor_runner`). `mtk_llama_executor_runner` is located inside `/cmake-android-out/examples/mediatek`
  - Question: should `executorch_runner` ever be used? If not, maybe not generate it?
Mention that they should use `mtk_llama_executor_runner` for llama models

**Comment by cccclai** [comment]
The changes in root cmake and the buffer allocator looks good! I marked them as resolved. The only one left is the destroy function and we're good then.

File: examples/mediatek/README.md
Comment (neuropilot-captain) [reply examples/mediatek/README.md:1]: > mtk_neuron seems to require a localhost connection. So verify that localhost connection is available (not blocked) before running the AoT script

Do you mean to guide users check their network condition before running this example?

Comment (Riandy) [reply examples/mediatek/README.md:1]: Yes. Give them some warning that a working localhost connection is required in AOT flow. When I tested the flow in Windows + Linux WSL (localhost is blocked) and I wasted 1.5 hours on the initial AoT flow without knowing that localhost is required.

File: backends/mediatek/runtime/NeuronBackend.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (cccclai) [reply backends/mediatek/runtime/NeuronBackend.cpp:86]: It seems like it's reverting to the old implementation, is it expected?

**Review by cccclai — APPROVED** [review]

**Comment by cccclai** [comment]
Many thanks for MTK team for making the contribution. I'll go ahead to merge it. Need to wait for the CI signal a bit.

**Comment by cccclai** [comment]
Looks like there are some lint errors. Could you address them? Here is the failing job https://github.com/pytorch/executorch/actions/runs/10296603326/job/28529110807?pr=3571

**Review by mergennachin — APPROVED** [review]

File: backends/mediatek/scripts/README.md
Comment (neuropilot-captain) [reply backends/mediatek/scripts/README.md:1]: No, we should remove in the future
